pkg/manager/impl/executions/quality_of_service.go

```diff
@@ -0,0 +1,164 @@
+	workflowIdentifier := input.Workflow.Id
+
+	var qualityOfServiceTier core.QualityOfService_Tier
+	if input.ExecutionCreateRequest.Spec.QualityOfService != nil {
+		if input.ExecutionCreateRequest.Spec.QualityOfService.GetSpec() != nil {
+			duration, err := ptypes.Duration(input.ExecutionCreateRequest.Spec.QualityOfService.GetSpec().QueueingBudget)
+			if err != nil {
+				return QualityOfServiceSpec{}, errors.NewFlyteAdminErrorf(codes.InvalidArgument,
+					"Invalid custom quality of service set in create execution request [%s/%s/%s], failed to parse duration [%v] with: %v",
+					input.ExecutionCreateRequest.Project, input.ExecutionCreateRequest.Domain,
+					input.ExecutionCreateRequest.Name,
+					input.ExecutionCreateRequest.Spec.QualityOfService.GetSpec().QueueingBudget, err)
+			}
+			return QualityOfServiceSpec{
+				QueuingBudget: duration,
+			}, nil
+		}
+		qualityOfServiceTier = input.ExecutionCreateRequest.Spec.QualityOfService.GetTier()
+	} else if input.LaunchPlan.Spec.QualityOfService != nil {
+		if input.LaunchPlan.Spec.QualityOfService.GetSpec() != nil {
+			duration, err := ptypes.Duration(input.LaunchPlan.Spec.QualityOfService.GetSpec().QueueingBudget)
+			if err != nil {
+				return QualityOfServiceSpec{}, errors.NewFlyteAdminErrorf(codes.InvalidArgument,
+					"Invalid custom quality of service set in launch plan [%v], failed to parse duration [%v] with: %v",
+					input.LaunchPlan.Id,
+					input.ExecutionCreateRequest.Spec.QualityOfService.GetSpec().QueueingBudget, err)
+			}
+			return QualityOfServiceSpec{
+				QueuingBudget: duration,
+			}, nil
+		}
+		qualityOfServiceTier = input.LaunchPlan.Spec.QualityOfService.GetTier()
+	} else if input.Workflow.Closure.CompiledWorkflow.Primary.Template.Metadata != nil &&
+		input.Workflow.Closure.CompiledWorkflow.Primary.Template.Metadata.QualityOfService != nil {
+		if input.Workflow.Closure.CompiledWorkflow.Primary.Template.Metadata.QualityOfService.GetSpec() != nil {
+			duration, err := ptypes.Duration(input.Workflow.Closure.CompiledWorkflow.Primary.Template.Metadata.QualityOfService.
+				GetSpec().QueueingBudget)
+			if err != nil {
+				return QualityOfServiceSpec{}, errors.NewFlyteAdminErrorf(codes.InvalidArgument,
+					"Invalid custom quality of service set in workflow [%v], failed to parse duration [%v] with: %v",
+					workflowIdentifier,
+					input.ExecutionCreateRequest.Spec.QualityOfService.GetSpec().QueueingBudget, err)
+			}
+			return QualityOfServiceSpec{
+				QueuingBudget: duration,
+			}, nil
+		}
+		qualityOfServiceTier = input.Workflow.Closure.CompiledWorkflow.Primary.Template.Metadata.QualityOfService.GetTier()
+	}
+
+	// If nothing in the hierarchy has set the quality of service, see if an override exists in the matchable attributes
+	// resource table.
+	resource, err := q.resourceManager.GetResource(ctx, interfaces.ResourceRequest{
```

Comment:
Should we only do this if qualityOfServiceTier == UNDEFINED ?

Reply:
good call, thanks! refactored this before opening the PR and missed the check here